chore(deps): update dependencies

iiroj opened the pull request:
This PR updates all top-level dependencies, expect for [log-symbols@5.0.0](https://github.com/sindresorhus/log-symbols/releases/tag/v5.0.0) which is now ESM-only and thus won't work in lint-staged until it is also converted to ESM.

What do you think, @okonet?

iiroj commented:
I rebased this and updated again, but it seems even more of @sindresorhus's packages have migrated to ESM and thus can no longer be updated.

I was able to migrate _lint-staged_ itself to ESM pretty easily, but the jest tests seem a bit more complicated. Some tests seems to fail because they "cannot use import statement outside a module", so it's probably related to Babel.

What do you think, @okonet?

okonet commented:
Sorry but I don't have an opinion on that one. I'm not sure why we need to migrate to ESM since lint-staged can't be used as a dependency.

iiroj commented:
To keep dependencies up-to-date, we need to either stop updating ESM deps, replace them, or convert to ESM ourselves.

Lint-staged does expose a Node.js API:

```js
import lintStaged from "lint-staged";
```

Personally I'm fine with anything, although if we stay at CommonJS, it might make sense to inline some simpler dependencies into the project directly.

okonet commented:
Hmm, yeah I forgot about providing the Node API :) 

I'm fine with everything too but I see you point of being up to date i.e. for security reasons. Can we ping someone from Jest / Babel to help us with this?

iiroj commented:
@okonet let's release all the updates we can do without changing anything, and then revisit once we know which dependencies can no longer be updated, and how the Node.js version requirements would change if we were to go to ESM.

I need to rebase this branch.

iiroj commented:
I rebased this and added some optimizations to reduce dependency sizes. See these commits: https://github.com/okonet/lint-staged/pull/1003/commits/5240c263e19ffa645a37bc16564d1f53d8f61e90 https://github.com/okonet/lint-staged/pull/1003/commits/4de4cdaea1e50c73ca01261bb8dcbd48287ec1ae

iiroj commented:
Ping @okonet this renovation got a little out of hand, and I also updated all the GitHub CI Action runners, as well as replace `yarn` with `npm`. Since the new [setup-node](https://github.com/actions/setup-node) supports caching out-of-box, using `npm` should be fine for performance.

Thoughts?

okonet commented:
LGTM

iiroj commented:
@okonet let's merge this as `feat` so that we can skip the buggy `11.1.3` and `11.1.4` published into `lint-staged@next` from here: https://github.com/okonet/lint-staged/pull/1022

sokra commented on package.json:
```diff
@@ -27,38 +27,38 @@
-    "debug": "^4.3.1",
+    "cli-truncate": "2.1.0",
+    "colorette": "^1.4.0",
+    "commander": "^8.2.0",
```
This dependency upgrade bumps the minimum node.js version from 10 to 12. So maybe that should be considered as breaking change instead?

okonet replied:
It for sure should have been! Thanks for pointing this out. cc @iiroj

iiroj replied:
Oops! Maybe a `fix` to downgrade? Or should we bump our version now that 16 is in LTS and 10 is getting old?

okonet replied:
I'd bump and release as a major TBH

iiroj replied:
I figured we could go all the way and convert _lint-staged_ to ESM: https://github.com/okonet/lint-staged/pull/1038

iiroj replied:
Actually, we already require Node.js 12.13, we just don't enforce it via the `engines.node` key:

https://github.com/okonet/lint-staged/blob/e035b80e39da355da57c02db6565b55271ab1afa/bin/lint-staged.js#L20
